lisp/x/profiler/opentelemetry_annotator.go

```diff
@@ -75,41 +72,27 @@ func contextTracer(ctx context.Context) trace.Tracer {
 	return otel.GetTracerProvider().Tracer(tracerName)
 }
 
-func (p *otelAnnotator) Start(function *lisp.LVal) {
+func (p *otelAnnotator) Start(function *lisp.LVal) func() {
 	if !p.enabled {
-		return
+		return func() {}
 	}
+	oldContext := p.currentContext
 	switch function.Type {
 	case lisp.LInt, lisp.LString, lisp.LFloat, lisp.LBytes, lisp.LError, lisp.LArray, lisp.LQuote, lisp.LNative, lisp.LQSymbol, lisp.LSortMap:
 		// We don't need to profile these types. We could, but we're not that LISP :D
-		return
+		return func() {}
 	case lisp.LFun, lisp.LSymbol, lisp.LSExpr:
 		fName := fmt.Sprintf("%s:%s", function.FunData().Package, getFunNameFromFID(p.runtime, function.FunData().FID))
-		p.contexts.Push(p.currentContext)
 		p.currentContext, p.currentSpan = contextTracer(p.currentContext).Start(p.currentContext, fName)
```

Comment:
Would it be safer to just keep the context and span as locals that the returned function uses to avoid any kind of race?

Reply:
i believe they are still needed so that on the recursive calls it does the right nesting, but maybe i'm not understanding the suggestion